editors/code/src/main.ts

```diff
@@ -109,3 +99,130 @@ export async function deactivate() {
+        "Update", "Cancel"
+    );
+    if (updateOrCancel !== "Update") return;
+
+    const artifact = release.assets.find(artifact => artifact.name === "rust-analyzer.vsix");
+    assert(!!artifact, `Bad release: ${JSON.stringify(release)}`);
+
+    const dest = path.join(config.globalStoragePath, "rust-analyzer.vsix");
+    await download(artifact.browser_download_url, dest, "Downloading rust-analyzer extension");
+
+    await vscode.commands.executeCommand("workbench.extensions.installExtension", vscode.Uri.file(dest));
+    await fs.unlink(dest);
+
+    await state.updateReleaseId(release.id);
+    await state.updateLastCheck(now);
+    await vscode.commands.executeCommand("workbench.action.reloadWindow");
+    assert(false, "unreachable");
+}
+
+async function bootstrapServer(config: Config, state: PersistentState): Promise<string> {
+    const path = await getServer(config, state);
+    if (!path) {
+        throw new Error(
+            "Rust Analyzer Language Server is not available. " +
+            "Please, ensure its [proper installation](https://rust-analyzer.github.io/manual.html#installation)."
+        );
+    }
+
+    const res = spawnSync(path, ["--version"], { encoding: 'utf8' });
+    log.debug("Checked binary availability via --version", res);
+    log.debug(res, "--version output:", res.output);
+    if (res.status !== 0) {
+        throw new Error(
+            `Failed to execute ${path} --version`
+        );
+    }
+
+    return path;
+}
+
+async function getServer(config: Config, state: PersistentState): Promise<string | undefined> {
+    const explicitPath = process.env.__RA_LSP_SERVER_DEBUG ?? config.serverPath;
+    if (explicitPath) {
+        if (explicitPath.startsWith("~/")) {
+            return os.homedir() + explicitPath.slice("~".length);
+        }
+        return explicitPath;
+    };
+
+    let binaryName: string | undefined = undefined;
+    if (process.arch === "x64" || process.arch === "x32") {
+        if (process.platform === "linux") binaryName = "rust-analyzer-linux";
+        if (process.platform === "darwin") binaryName = "rust-analyzer-mac";
+        if (process.platform === "win32") binaryName = "rust-analyzer-windows.exe";
+    }
+    if (binaryName === undefined) {
+        vscode.window.showErrorMessage(
+            "Unfortunately we don't ship binaries for your platform yet. " +
+            "You need to manually clone rust-analyzer repository and " +
+            "run `cargo xtask install --server` to build the language server from sources. " +
+            "If you feel that your platform should be supported, please create an issue " +
+            "about that [here](https://github.com/rust-analyzer/rust-analyzer/issues) and we " +
+            "will consider it."
+        );
+        return undefined;
+    }
+
+    const dest = path.join(config.globalStoragePath, binaryName);
+    const exists = await fs.stat(dest).then(() => true, () => false);
+    if (!exists) {
```

Comment:
```suggestion
    if (!fs.existsSync(dest)) {
```

Reply:
FIY: the async `fs.exists()` version is deprecated...

Reply:
Oh, and for this we should amend the import: 
```ts
import { promises as fs, existsSync as doesFileExist } from "fs";
```

Reply:
I believe async stat is the more correct approach.